flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/GenericRecordAvroTypeInfo.java

```diff
@@ -104,10 +105,15 @@ public boolean canEqual(Object obj) {
     }
 
     private void writeObject(ObjectOutputStream oos) throws IOException {
-        oos.writeUTF(schema.toString());
+        byte[] schemaStrInBytes = schema.toString(false).getBytes(StandardCharsets.UTF_8);
```

Comment:
My main question is the compatibility. could TypeInfo object be stored in checkpoint state? if yes, this change will break the state restoring.

Reply:
What gets stored in the the checkpoint/savepoint is the type serializer snapshot. The TypeInfo would be part of the serialized job graph. AFAIK, that does not survive a restart?

Reply:
Good catch. I wasn't previously thinking about state restoring.

I think type serializer does need to be checkpointed and restored (code [link](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSerializerProvider.java#L99))

I'm less sure about how type info serialization is used as I didn't manage to find a code reference for its usage.

Reply:
@tweise @stevenzwu I think the avro serializer snapshot (which is used for de/ser operator state if the state include Avro record) already takes care of de/ser of large Avro schema: Code [link](https://github.com/apache/flink/blob/master/flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSerializerSnapshotTest.java#L155).

Since the serialization of avro serializer and typeinfo is only used when constructing edge and vertices of stream graph, maybe we don't need to take care of backward compatibility here.